### DISCUSSION TITLE
How is external address confirmed?

### DISCUSSION THREAD

**Discussion by Wiezzel** [libp2p/rust-libp2p#5099] [Q&A] [answered]
Using `rust-libp2p` version `0.53.2`, I have noticed that nodes do not automatically discover their public addresses. After running a node for quite a while I see lots of `NewExternalAddrCandidate` events, but never an `ExternalAddrConfirmed`. There are also incoming connections established using the node's public IP, so it seems that the external should be confirmed at this point, but the event never occurs. So my question is: what is the condition under which `ExternalAddrConfirmed` is emitted? Do I need to include some additional network behaviour or write some extra code for this to happen?

**Comment by dariusc93** [comment, marked as answer]
It comes down to what behaviour youre using or if youre manually adding the external address. Eg you can use autonat to determine your node external address (assuming your node is dialable and connected to another peer with the same protocol). You can use `libp2p-upnp`, which would report your ip address via `ExternalAddrConfirmed` if upnp is supported. You can also manually add it via `Swarm::add_external_address`, which would also emit the event to swarm.

**Reply by Wiezzel** [reply]
Okay, so it's up to the behaviour. Roger that.
I was expecting that maybe swarm does emit this event on its own sometimes. I mean, it would make sense to emit `ExternalAddrConfirmed` is another node dials this address and connects to ours, wouldn't it?

**Reply by thomaseizinger** [reply]
> I was expecting that maybe swarm does emit this event on its own sometimes. I mean, it would make sense to emit `ExternalAddrConfirmed` is another node dials this address and connects to ours, wouldn't it?

What you are describing is essential the AutoNAT protocol:

- Tell another node our addresses
- Have it dial the address
- If it works, consider it reachable

You can follow the development of the AutoNATv2 protocol here: https://github.com/umgefahren/rust-libp2p/pull/1. The spec is here: https://github.com/libp2p/specs/pull/538

As it turns out, it can be a bit tricky because you have to make sure you are not accidentally building a botnet :)
